### PR TITLE
refactor: cleanup database recovery (#3225)

### PR DESCRIPTION
**Pull request by tustvold**
In preparation for #3225 this PR does some minor cleanup to the existing WipePreservedCatalog and SkipReplay code paths to hopefully reduce the duplication.

This also finishes up the work started by @carols10cents in #2570 so that the API for Database is slightly more consistent

### CODE REVIEW

File: influxdb_iox/src/influxdb_ioxd/server_type/database/rpc/management.rs [hunk 1]
```diff
@@ -421,15 +421,8 @@ impl management_service_server::ManagementService for ManagementService {
         let tracker = self
             .server
             .wipe_preserved_catalog(&db_name)
-            .map_err(|e| match e {
```
Comment (tustvold) [inline influxdb_iox/src/influxdb_ioxd/server_type/database/rpc/management.rs:424]: This error mapping confused me, as I can't see  a way for server::wipe_preserved_catalog to return anything other than DatabaseNotFound or WipePreservedCatalog

File: server/src/lib.rs [hunk 2]
```diff
@@ -874,24 +874,14 @@ impl Server {
     ///
     /// The DB must exist in the server and have failed to load the catalog for this to work
     /// This is done to prevent race conditions between DB jobs and this command
-    pub fn wipe_preserved_catalog(
```
Comment (tustvold) [inline server/src/lib.rs:877]: I'm tempted to just remove this method from server in a future PR

File: influxdb_iox/src/influxdb_ioxd/server_type/database/rpc/error.rs [hunk 3]
```diff
@@ -1,5 +1,5 @@
 use generated_types::google::{
-    FieldViolation, InternalError, NotFound, PreconditionViolation, QuotaFailure,
+    AlreadyExists, FieldViolation, InternalError, NotFound, PreconditionViolation, QuotaFailure,
```
Comment (tustvold) [inline influxdb_iox/src/influxdb_ioxd/server_type/database/rpc/error.rs:2]: I wonder about defining these conversions as From implementations, to remove the need for all the map_err everywhere :thinking:

File: influxdb_iox/src/influxdb_ioxd/server_type/database/rpc/error.rs [hunk 4]
```diff
@@ -17,6 +17,12 @@ pub fn default_server_error_handler(error: server::Error) -> tonic::Status {
         Error::DatabaseNotInitialized { db_name } => {
             tonic::Status::unavailable(format!("Database ({}) is not yet initialized", db_name))
         }
+        Error::DatabaseAlreadyExists { db_name } => AlreadyExists {
```
Comment (tustvold) [inline influxdb_iox/src/influxdb_ioxd/server_type/database/rpc/error.rs:20]: Aside from more consistent error messages, using the `AlreadyExists` payload will embed the corresponding grpc details payload for clients to potentially decode

File: influxdb_iox/tests/end_to_end_cases/management_api.rs [hunk 5]
```diff
@@ -338,7 +338,7 @@ async fn test_create_get_update_release_claim_database() {
     let err = client.claim_database(released_uuid).await.unwrap_err();
     assert_contains!(
         err.to_string(),
-        format!("A database with the name `{}` already exists", db_name)
+        format!("Resource database/{} already exists", db_name)
```
Comment (tustvold) [inline influxdb_iox/tests/end_to_end_cases/management_api.rs:341]: This is a consequence of changing the default error mapping, this makes it consistent with the other `AlreadyExists` style errors, in addition to providing the grpc details payload should a client wish to decode it

File: influxdb_iox/src/influxdb_ioxd/server_type/database/rpc/error.rs (hunk 3 above)
Comment (crepererum) [reply influxdb_iox/src/influxdb_ioxd/server_type/database/rpc/error.rs:2]: I think the issue is:

1. we don't map all errors
2. For the ones we don't map, we convert them to an opaque "internal error" and create a log statement for debugging. IMHO `From` should be free from side effects, so logging isn't really an option.
3. Sometimes we convert to google errors, sometimes we directly create a `tonic::Status`.

Overall I feel that  the error handling handling could also be improved by not having these overlarge error types w/ ton of variants :/

File: server/src/database.rs [hunk 6]
```diff
@@ -28,22 +29,30 @@ use std::{future::Future, sync::Arc, time::Duration};
 use time::Time;
 use tokio::{sync::Notify, task::JoinError};
 use tokio_util::sync::CancellationToken;
+use tracker::{TaskTracker, TrackedFutureExt};
 use uuid::Uuid;
 
+/// Matches an error [`DatabaseState`] and clones the contained state
+macro_rules! error_state {
+    ($s:expr, $transition: literal, $variant:ident) => {
+        match &**$s.shared.state.read() {
```
Comment (crepererum) [inline server/src/database.rs:38]: `&**$`

Matrix

File: server/src/lib.rs (hunk 2 above)
Comment (crepererum) [reply server/src/lib.rs:877]: I agree. Most of the method on `Server` should be removed. API users should either user `Database` or `Db` directly.

File: server/src/database.rs [hunk 7]
```diff
@@ -487,47 +496,40 @@ impl Database {
...
-                _ => {
-                    return InvalidState {
-                        db_name,
-                        state: state.state_code(),
-                        transition: "WipePreservedCatalog",
-                    }
-                    .fail()
-                }
-            };
+    pub async fn wipe_preserved_catalog(&self) -> Result<TaskTracker<Job>, Error> {
```
Comment (crepererum) [inline server/src/database.rs:499]: why is this `async` now?

File: server/src/database.rs [hunk 8]
```diff
@@ -487,47 +496,40 @@ impl Database {
...
+        let registry = self.shared.application.job_registry();
+        let (tracker, registration) = registry.register(Job::WipePreservedCatalog {
+            db_name: Arc::from(db_name.as_str()),
+        });
 
         let shared = Arc::clone(&self.shared);
 
-        Ok(async move {
-            let db_name = &shared.config.name;
+        tokio::spawn(
```
Comment (crepererum) [inline server/src/database.rs:513]: Who tracks this subtask and can stop it? I'm somewhat allergic to untracked `tokio::spawn` calls.

File: server/src/database.rs (hunk 7 above)
Comment (tustvold) [reply server/src/database.rs:499]: It allows removing the `TransitionInProgress` error variant. I'm not sure how I feel about this, I was just going for consistency with the other methods - I personally am not a massive fan of yet more async :laughing:

File: server/src/database.rs (hunk 8 above)
Comment (tustvold) [reply server/src/database.rs:513]: The tracker registration can stop it, this is just the tokio::spawn that used to be in server moved into Database

File: server/src/database.rs (hunk 7 above)
Comment (crepererum) [reply server/src/database.rs:499]: I don't understand how adding an `async` to a function that never `await`s allows you to remove an enum variant :thinking:

Comment (tustvold) [reply server/src/database.rs:499]: There is a `handle.await` two lines down?

File: server/src/database.rs (hunk 8 above)
Comment (crepererum) [reply server/src/database.rs:513]: Uuh, I will probably ban `tokio::spawn` at one point and allow this call only at places that opt-in, e.g. this `spawn` call could be done somewhere in the task registration. But anyway, that's not your concern now.

**Review by crepererum — APPROVED** [review]

File: server/src/database.rs (hunk 8 above)
Comment (tustvold) [reply server/src/database.rs:513]: I would not object to such a change

**Pull request merged**